### PR TITLE
Create loader for Building Scene Layer JSON structure.

### PR DESCRIPTION
**Pull request by dryabinin-actionengine**

### CODE REVIEW

File: modules/i3s/src/i3s-building-scene-layer-loader.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  const layers: BuildingSceneSublayer[] = [];
+  const layer0 = JSON.parse(new TextDecoder().decode(data));
+  const {sublayers} = layer0;
+  const url = context?.url;
+
+  if (!url) {
+    throw new Error('Url is not provided');
+  }
+
+  parseSublayersTree(sublayers, url, layers);
```
Comment (belom88) [inline modules/i3s/src/i3s-building-scene-layer-loader.ts:38]: Lets create return for this function and create 'layers' array inside there

File: modules/i3s/src/i3s-building-scene-layer-loader.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+  return {
+    header: layer0,
+    sublayers: layers
```
Comment (belom88) [inline modules/i3s/src/i3s-building-scene-layer-loader.ts:42]:
```suggestion
    sublayers: parserSublayersTree(sublayers, url)
```

**Review by belom88 — APPROVED** [review]
I see that you've made `const layers: BuildingSceneSublayer[] = [];` this way because of recursion. But I believe there is a way to make the recursive function with return.

File: modules/i3s/src/i3s-building-scene-layer-loader.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        layerType,
+        ...rest
+      });
+    }
+
+    if (subLayer?.sublayers?.length) {
+      parseSublayersTree(subLayer.sublayers, url, layers);
+    }
+  }
+}
```
Comment (belom88) [inline modules/i3s/src/i3s-building-scene-layer-loader.ts:71]:
```suggestion
function parseSublayersTree(
  sublayers: BuildingSceneSublayer[],
  url: string
): BuildingSceneSublayer[] {
  let layers: BuildingSceneSublayer[] = [];
  for (let index = 0; index < sublayers.length; index++) {
    const subLayer = sublayers[index];
    const {id, layerType, ...rest} = subLayer;

    // Add support only for 3DObject layer type for I3S purposes.
    if (layerType === OBJECT_3D_LAYER_TYPE) {
      const sublayerUrl = `${url}/sublayers/${id}`;

      layers.push({
        url: sublayerUrl,
        id,
        layerType,
        ...rest
      });
    }

    if (subLayer?.sublayers?.length) {
      layers = [...layers, ...parseSublayersTree(subLayer.sublayers, url)];
    }
  }
  return layers;
}
```

File: modules/i3s/src/i3s-building-scene-layer-loader.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+  parseSublayersTree(sublayers, url, layers);
+
+  return {
+    header: layer0,
+    sublayers: layers
+  };
+}
+
+function parseSublayersTree(
```
Comment (ibgreen) [inline modules/i3s/src/i3s-building-scene-layer-loader.ts:46]: Nits:
- JSDoc on the functions?
- Move long parsing code out of the loader file into `lib` folder?

**Review by ibgreen — APPROVED** [review]
Should we do a pass on the  is3 module / I3SLoader docs and describe what layers etc we actually support?

File: modules/i3s/src/i3s-building-scene-layer-loader.ts (hunk 1 above)
Comment (dryabinin-actionengine) [reply modules/i3s/src/i3s-building-scene-layer-loader.ts]: Done

File: modules/i3s/src/i3s-building-scene-layer-loader.ts (hunk 2 above)
Comment (dryabinin-actionengine) [reply modules/i3s/src/i3s-building-scene-layer-loader.ts]: Done

File: modules/i3s/src/i3s-building-scene-layer-loader.ts (hunk 3 above)
Comment (dryabinin-actionengine) [reply modules/i3s/src/i3s-building-scene-layer-loader.ts]: Done

File: modules/i3s/src/i3s-building-scene-layer-loader.ts (hunk 4 above)
Comment (dryabinin-actionengine) [reply modules/i3s/src/i3s-building-scene-layer-loader.ts]: Done

**Comment by dryabinin-actionengine** [comment]
@ibgreen Will add docs in next PR

**Pull request merged**